pkg/authorization/authorizer/accessrestriction/accessrestriction.go

```diff
@@ -0,0 +1,175 @@
+
+	"github.com/golang/glog"
+)
+
+func NewAuthorizer(accessRestrictionLister authorizationlister.AccessRestrictionLister, userLister userlisterv1.UserLister, groupLister userlisterv1.GroupLister) authorizer.Authorizer {
+	return &accessRestrictionAuthorizer{
+		accessRestrictionLister: accessRestrictionLister,
+		userLister:              userLister,
+		groupLister:             groupLister,
+	}
+}
+
+type accessRestrictionAuthorizer struct {
+	accessRestrictionLister authorizationlister.AccessRestrictionLister
+	userLister              userlisterv1.UserLister
+	groupLister             userlisterv1.GroupLister
+}
+
+func (a *accessRestrictionAuthorizer) Authorize(requestAttributes authorizer.Attributes) (authorizer.Decision, string, error) {
+	accessRestrictions, err := a.accessRestrictionLister.List(labels.Everything())
+	if err != nil {
+		// fail closed (but this should never happen because it means some static generated code is broken)
+		return authorizer.DecisionDeny, "cannot determine access restrictions", err
+	}
+
+	// check all access restrictions and only short circuit on affirmative deny
+	for _, accessRestriction := range accessRestrictions {
+		// does this access restriction match the given request attributes
+		if matches(accessRestriction, requestAttributes) {
+			// it does match, meaning we need to check if it denies the request
+			if !a.allowed(accessRestriction, requestAttributes.GetUser()) {
+				// deny the request because it is not allowed by the current access restriction
+				// the reason is opaque because normal users have no visibility into access restriction objects
+				glog.V(2).Infof("access restriction %#v denied request attributes %#v for user %#v", accessRestriction, requestAttributes, requestAttributes.GetUser())
+				return authorizer.DecisionDeny, "denied by access restriction", nil
+			}
+			glog.V(4).Infof("access restriction %#v matched but did not deny request attributes %#v for user %#v", accessRestriction, requestAttributes, requestAttributes.GetUser())
+		}
+	}
+
+	// no access restriction matched or denied this request, so we state that we have no opinion
+	// the reason must be blank, otherwise we would spam all RBAC denies with it (which is generally not useful)
+	return authorizer.DecisionNoOpinion, "", nil
+}
+
+func matches(accessRestriction *authorization.AccessRestriction, requestAttributes authorizer.Attributes) bool {
+	if len(accessRestriction.Spec.MatchAttributes) == 0 {
+		return true // fail closed (but validation prevents this)
+	}
+	return rbac.RulesAllow(requestAttributes, accessRestriction.Spec.MatchAttributes...)
+}
+
+func (a *accessRestrictionAuthorizer) allowed(accessRestriction *authorization.AccessRestriction, user user.Info) bool {
+	s := accessRestriction.Spec
+	isWhitelist := len(s.AllowedSubjects) != 0 && len(s.DeniedSubjects) == 0
+	isBlacklist := len(s.DeniedSubjects) != 0 && len(s.AllowedSubjects) == 0
+
+	switch {
+	case isWhitelist:
```

Comment:
this is weird.  Why wouldn't I simply not deny them at all?  Given the list without any ordering guarantees it's not like this is reliably skipping other rules.